stl/inc/syncstream

```diff
@@ -0,0 +1,312 @@
+public:
+    using int_type       = typename _Traits::int_type;
+    using pos_type       = typename _Traits::pos_type;
+    using off_type       = typename _Traits::off_type;
+    using allocator_type = _Alloc;
+    using streambuf_type = basic_streambuf<_Elem, _Traits>;
+
+    using pointer   = typename allocator_traits<_Alloc>::pointer;
+    using size_type = typename allocator_traits<_Alloc>::size_type;
+
+    basic_syncbuf() : basic_syncbuf(nullptr){};
+
+    explicit basic_syncbuf(streambuf_type* _StrBuf) : basic_syncbuf(_StrBuf, _Alloc{}) {}
+
+    basic_syncbuf(streambuf_type* _StrBuf, const _Alloc& _Al) : _Wrapped(_StrBuf), _Al(_Al) {
+        if (_Wrapped != nullptr) {
+            _Add_mutex_for_instance_or_increment(static_cast<void*>(_Wrapped));
+        }
+        _Init();
+    }
+
+    basic_syncbuf(basic_syncbuf&& _Right) {
+        _Al = _STD move(_Right._Al);
+        _Assign_rv(_STD move(_Right));
+    }
+
+    ~basic_syncbuf() {
+        _Emit();
+        _Tidy();
+    }
+
+    basic_syncbuf& operator=(basic_syncbuf&& _Right) {
+        _Emit();
+        _Move_assign(_STD move(_Right), _Choose_pocma<_Alloc>{});
+        return *this;
+    }
+
+    void swap(basic_syncbuf& _Right) {
+        if (this != _STD addressof(_Right)) {
+            _Pocs(_Getal(), _Right._Getal());
+            _Swap_except_al(_Right);
+        }
+    }
+
+    bool emit() {
+        if (_Wrapped == nullptr) {
+            return false;
+        }
+
+        bool _Result                 = true;
+        const size_type _Data_size   = _Get_data_size();
+        const pointer _Begin_seq_ptr = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        if (_Data_size > 0 || _Record_sync) {
+            scoped_lock _Guard(_Get_mutex_for_instance(static_cast<void*>(_Wrapped)));
+
+            if (_Data_size > 0 && _Data_size != _Wrapped->sputn(_Begin_seq_ptr, _Data_size)) {
+                _Result = false;
+            }
+
+            if (_Record_sync) {
+                if (_Wrapped->pubsync() == -1) {
+                    _Result = false;
+                }
+            }
+        }
+        _Record_sync = false;
+        streambuf_type::setp(
+            _Begin_seq_ptr, pointer_traits<pointer>::pointer_to(*streambuf_type::epptr())); // reset written data
+        return _Result;
+    }
+
+    streambuf_type* get_wrapped() const noexcept {
+        return _Wrapped;
+    }
+
+    allocator_type get_allocator() const noexcept {
+        return _Getal();
+    }
+
+    void set_emit_on_sync(bool _Val) noexcept {
+        _Emit_on_sync = _Val;
+    }
+
+protected:
+    int sync() override {
+        _Record_sync = true;
+
+        if (_Emit_on_sync) {
+            if (!emit()) {
+                return -1;
+            }
+        }
+        return 0;
+    }
+
+    int_type overflow(int_type _Current_elem) override {
+        if (_Wrapped == nullptr)
+            return _Traits::eof();

```

Comment:
We conventionally use braces for all control flow; occurs elsewhere. (This helps avoid mistakes during maintenance, and we feel that the consistency is worth the additional line.)